### PR TITLE
Best effort to delete dataset version if named blob upload failed

### PR DESCRIPTION
**Pull request by SophieGuo410**
We add the dataset version before uploading the named blob in order to support LATEST version control.
If named blob upload failed(including upload blob and update named blob db), we will take the best effort to delete the specific dataset version.

### CODE REVIEW

**Review by JingQianCloud — COMMENTED** [review]
NamedBlobPutHandler has a Callback Chain with multiple steps.
Looks like the PR is carefully picking for each step, shall we use the "finalCallback" or the "deleteDatasetCallback"?
Is there any harm to always call "deleteDatasetCallback"?

If so, we probably can only change one line
this.finalCallback = deleteDatasetVersionIfUploadFailedCallBack(finalCallback);

**Comment by SophieGuo410** [comment]
> NamedBlobPutHandler has a Callback Chain with multiple steps. Looks like the PR is carefully picking for each step, shall we use the "finalCallback" or the "deleteDatasetCallback"? Is there any harm to always call "deleteDatasetCallback"?
> 
> If so, we probably can only change one line this.finalCallback = deleteDatasetVersionIfUploadFailedCallBack(finalCallback);

In NamedBlobPutHandler, we might call finalCallback for both successAction and failureCallback, you can refer details in buildCallback. So I just replace the finalCallback with deleteDatasetCallback when named blob upload or router.putBlob related function failed.

File: ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobPutHandler.java [hunk 1]
```diff
@@ -321,7 +323,7 @@ private Callback<Void> routerTtlUpdateCallback(BlobInfo blobInfo, String blobId)
           deleteDatasetVersionOutOfRetentionCount();
         }
         securityService.processResponse(restRequest, restResponseChannel, blobInfo, securityProcessResponseCallback());
-      }, uri, LOGGER, finalCallback);
+      }, uri, LOGGER, deleteDatasetCallback);
     }
```
Comment (JingQianCloud) [inline ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobPutHandler.java:327]: For the last step, securityProcessResponseCallback, shall we change the failure callback to deleteDatasetCallback as well?

I feel for any buildCallback after the PutBlob step, we should replace the failure callback to deleteDatasetCallback.

```   
private Callback<Void> securityProcessResponseCallback() {
      return buildCallback(frontendMetrics.putBlobSecurityProcessResponseMetrics,
          securityCheckResult -> finalCallback.onCompletion(null, null), restRequest.getUri(), LOGGER, finalCallback);
```

Comment (SophieGuo410) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobPutHandler.java:327]: This step is to sumbit the response. the named blob and regular blob already in our service. So I feel like no need to rollback if only the sumbit response failed. Instead, client will take the best effort to manually delete this version.

**Review by JingQianCloud — APPROVED** [review]

**Review by justinlin-linkedin — APPROVED** [review]

**Pull request merged**